File: mobile/ios/Security/RASPManager.swift
import Foundation
import Security
import UIKit

class RASPManager {
    static let shared = RASPManager()
    
    private var isMonitoring = false
    private var securityTimer: Timer?
    private var threatCount = 0
    private var lastThreatTime: Date?
    
    private init() {}
    
    // MARK: - Main Methods
    
    /// Запуск мониторинга безопасности
    func startMonitoring() {
        guard !isMonitoring else { return }
        
        isMonitoring = true
        threatCount = 0
        lastThreatTime = nil
        
        print("🛡️ RASP: Starting security monitoring...")
        startSecurityChecks()
    }
    
    /// Остановка мониторинга
    func stopMonitoring() {
        guard isMonitoring else { return }
        
        isMonitoring = false
        securityTimer?.invalidate()
        securityTimer = nil
        
        print("🛡️ RASP: Security monitoring stopped")
    }
    
    /// Получить статус мониторинга
    func getMonitoringStatus() -> RASPStatus {
        return RASPStatus(
            isMonitoring: isMonitoring,
            threatCount: threatCount,
            lastThreatTime: lastThreatTime,
            securityLevel: determineSecurityLevel()
        )
    }
    
    // MARK: - Security Checks
    
    private func startSecurityChecks() {
        securityTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { _ in
            self.performSecurityChecks()
        }
    }
    
    private func performSecurityChecks() {
        // 1. Проверка отладки
        if isBeingDebugged() {
            handleSecurityThreat(.debugging)
        }
        
        // 2. Проверка модификации кода
        if isCodeModified() {
            handleSecurityThreat(.codeModification)
        }
        
        // 3. Проверка инъекций
        if isCodeInjected() {
            handleSecurityThreat(.codeInjection)
        }
        
        // 4. Проверка эмуляции
        if isRunningOnEmulator() {
            handleSecurityThreat(.emulation)
        }
        
        // 5. Проверка hooking
        if isHooked() {
            handleSecurityThreat(.hooking)
        }
        
        // 6. Проверка памяти
        if isMemoryTampered() {
            handleSecurityThreat(.memoryTampering)
        }
        
        // 7. Проверка целостности
        if isIntegrityCompromised() {
            handleSecurityThreat(.integrityViolation)
        }
    }
    
    // MARK: - Detection Methods
    
    // 1. Проверка отладки
    private func isBeingDebugged() -> Bool {
        // Проверка через ptrace
        let result = ptrace(PT_DENY_ATTACH, 0, 0, 0)
        if result == -1 {
            return true
        }
        
        // Проверка через sysctl
        var info = kinfo_proc()
        var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
        var size = MemoryLayout<kinfo_proc>.stride
        
        let result2 = sysctl(&mib, u_int(mib.count), &info, &size, nil, 0)
        if result2 == 0 {
            return (info.kp_proc.p_flag & P_TRACED) != 0
        }
        
        return false
    }
    
    // 2. Проверка модификации кода
    private func isCodeModified() -> Bool {
        // Проверка подписи приложения
        guard let bundlePath = Bundle.main.bundlePath.cString(using: .utf8) else {
            return true
        }
        
        let code = SecStaticCodeCreateWithPath(URL(fileURLWithPath: Bundle.main.bundlePath) as CFURL, [], nil)
        guard let staticCode = code else {
            return true
        }
        
        let result = SecStaticCodeCheckValidity(staticCode, [], nil)
        return result != errSecSuccess
    }
    
    // 3. Проверка инъекций
    private func isCodeInjected() -> Bool {
        // Проверка на подозрительные библиотеки
        let suspiciousLibraries = [
            "MobileSubstrate",
            "libsubstrate",
            "libhooker",
            "libsubstitute",
            "cynject",
            "libcynject"
        ]
        
        for library in suspiciousLibraries {
            if dlopen(library, RTLD_NOW) != nil {
                return true
            }
        }
        
        return false
    }
    
    // 4. Проверка эмуляции
    private func isRunningOnEmulator() -> Bool {
        // Проверка архитектуры
        #if targetEnvironment(simulator)
        return true
        #else
        return false
        #endif
    }
    
    // 5. Проверка hooking
    private func isHooked() -> Bool {
        // Проверка на подозрительные символы
        let suspiciousSymbols = [
            "MSHookFunction",
            "MSHookMessageEx",
            "MSFindSymbol",
            "MSGetImageByName"
        ]
        
        for symbol in suspiciousSymbols {
            if dlsym(dlopen(nil, RTLD_NOW), symbol) != nil {
                return true
            }
        }
        
        return false
    }
    
    // 6. Проверка памяти
    private func isMemoryTampered() -> Bool {
        // Проверка на подозрительные паттерны в памяти
        let memoryAddress = UnsafeRawPointer(bitPattern: 0x100000000)
        guard let address = memoryAddress else { return false }
        
        // Простая проверка на известные паттерны
        let pattern = "HOOK".data(using: .utf8)!
        let memoryData = Data(bytes: address, count: pattern.count)
        
        return memoryData == pattern
    }
    
    // 7. Проверка целостности
    private func isIntegrityCompromised() -> Bool {
        // Проверка на изменения в критических секциях
        let criticalSections = [
            "ALADDINMobile",
            "SecurityManager",
            "CertificatePinningManager"
        ]
        
        for section in criticalSections {
            if let symbol = dlsym(dlopen(nil, RTLD_NOW), section) {
                // Проверяем, что символ не был изменен
                let originalValue = dlsym(dlopen(nil, RTLD_NOW), section)
                if symbol != originalValue {
                    return true
                }
            }
        }
        
        return false
    }
    
    // MARK: - Threat Handling
    
    private func handleSecurityThreat(_ threat: SecurityThreat) {
        threatCount += 1
        lastThreatTime = Date()
        
        print("🚨 RASP Threat Detected: \(threat.rawValue)")
        
        // Отправляем уведомление
        NotificationCenter.default.post(
            name: .raspThreatDetected,
            object: nil,
            userInfo: ["threat": threat.rawValue, "count": threatCount]
        )
        
        // Выполняем защитные действия
        executeDefensiveActions(for: threat)
    }
    
    private func executeDefensiveActions(for threat: SecurityThreat) {
        switch threat {
        case .debugging:
            // Завершаем приложение при отладке
            exit(1)
            
        case .codeModification:
            // Блокируем подозрительные функции
            disableSensitiveFeatures()
            
        case .codeInjection:
            // Очищаем память
            clearSensitiveData()
            
        case .emulation:
            // Ограничиваем функциональность
            limitFunctionality()
            
        case .hooking:
            // Перезапускаем критические компоненты
            restartCriticalComponents()
            
        case .memoryTampering:
            // Очищаем и перезагружаем данные
            reloadSecurityData()
            
        case .integrityViolation:
            // Активируем максимальную защиту
            activateMaximumSecurity()
        }
    }
    
    // MARK: - Defensive Actions
    
    private func disableSensitiveFeatures() {
        print("🛡️ RASP: Disabling sensitive features")
        // В реальном приложении здесь была бы логика отключения функций
    }
    
    private func clearSensitiveData() {
        print("🛡️ RASP: Clearing sensitive data")
        // В реальном приложении здесь была бы очистка данных
    }
    
    private func limitFunctionality() {
        print("🛡️ RASP: Limiting functionality")
        // В реальном приложении здесь было бы ограничение функций
    }
    
    private func restartCriticalComponents() {
        print("🛡️ RASP: Restarting critical components")
        // В реальном приложении здесь был бы перезапуск компонентов
    }
    
    private func reloadSecurityData() {
        print("🛡️ RASP: Reloading security data")
        // В реальном приложении здесь была бы перезагрузка данных
    }
    
    private func activateMaximumSecurity() {
        print("🛡️ RASP: Activating maximum security")
        // В реальном приложении здесь была бы активация максимальной защиты
    }
    
    // MARK: - Helper Methods
    
    private func determineSecurityLevel() -> SecurityLevel {
        if threatCount == 0 {
            return .high
        } else if threatCount < 3 {
            return .medium
        } else {
            return .low
        }
    }
    
    /// Получить детальный отчет
    func getDetailedReport() -> String {
        let status = getMonitoringStatus()
        
        var report = "🛡️ RASP Security Report\n"
        report += "======================\n"
        report += "Monitoring: \(status.isMonitoring ? "✅ ACTIVE" : "❌ INACTIVE")\n"
        report += "Threats Detected: \(status.threatCount)\n"
        report += "Last Threat: \(status.lastThreatTime?.description ?? "None")\n"
        report += "Security Level: \(status.securityLevel.rawValue)\n"
        
        return report
    }
}

// MARK: - Data Models

struct RASPStatus {
    let isMonitoring: Bool
    let threatCount: Int
    let lastThreatTime: Date?
    let securityLevel: SecurityLevel
}

enum SecurityThreat: String, CaseIterable {
    case debugging = "Debugging"
    case codeModification = "Code Modification"
    case codeInjection = "Code Injection"
    case emulation = "Emulation"
    case hooking = "Hooking"
    case memoryTampering = "Memory Tampering"
    case integrityViolation = "Integrity Violation"
}

enum SecurityLevel: String, CaseIterable {
    case low = "Low"
    case medium = "Medium"
    case high = "High"
}

// MARK: - Notifications

extension Notification.Name {
    static let raspThreatDetected = Notification.Name("RASPThreatDetected")
}

File: mobile/android/Security/RASPManager.kt
package com.aladdin.mobile.security

import android.content.Context
import android.os.Debug
import android.os.Process
import android.util.Log
import java.io.File
import java.io.IOException
import java.lang.reflect.Method
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RASPManager @Inject constructor(
    private val context: Context
) {
    
    private var isMonitoring = false
    private var threatCount = 0
    private var lastThreatTime: Long = 0
    private var securityTimer: android.os.Handler? = null
    
    // MARK: - Main Methods
    
    /**
     * Запуск мониторинга безопасности
     */
    fun startMonitoring() {
        if (isMonitoring) return
        
        isMonitoring = true
        threatCount = 0
        lastThreatTime = 0
        
        Log.i("RASP", "Starting security monitoring...")
        startSecurityChecks()
    }
    
    /**
     * Остановка мониторинга
     */
    fun stopMonitoring() {
        if (!isMonitoring) return
        
        isMonitoring = false
        securityTimer?.removeCallbacksAndMessages(null)
        securityTimer = null
        
        Log.i("RASP", "Security monitoring stopped")
    }
    
    /**
     * Получить статус мониторинга
     */
    fun getMonitoringStatus(): RASPStatus {
        return RASPStatus(
            isMonitoring = isMonitoring,
            threatCount = threatCount,
            lastThreatTime = lastThreatTime,
            securityLevel = determineSecurityLevel()
        )
    }
    
    // MARK: - Security Checks
    
    private fun startSecurityChecks() {
        securityTimer = android.os.Handler(android.os.Looper.getMainLooper())
        securityTimer?.postDelayed(object : Runnable {
            override fun run() {
                if (isMonitoring) {
                    performSecurityChecks()
                    securityTimer?.postDelayed(this, 1000) // Проверка каждую секунду
                }
            }
        }, 1000)
    }
    
    private fun performSecurityChecks() {
        // 1. Проверка отладки
        if (isBeingDebugged()) {
            handleSecurityThreat(SecurityThreat.DEBUGGING)
        }
        
        // 2. Проверка модификации кода
        if (isCodeModified()) {
            handleSecurityThreat(SecurityThreat.CODE_MODIFICATION)
        }
        
        // 3. Проверка инъекций
        if (isCodeInjected()) {
            handleSecurityThreat(SecurityThreat.CODE_INJECTION)
        }
        
        // 4. Проверка эмуляции
        if (isRunningOnEmulator()) {
            handleSecurityThreat(SecurityThreat.EMULATION)
        }
        
        // 5. Проверка hooking
        if (isHooked()) {
            handleSecurityThreat(SecurityThreat.HOOKING)
        }
        
        // 6. Проверка памяти
        if (isMemoryTampered()) {
            handleSecurityThreat(SecurityThreat.MEMORY_TAMPERING)
        }
        
        // 7. Проверка целостности
        if (isIntegrityCompromised()) {
            handleSecurityThreat(SecurityThreat.INTEGRITY_VIOLATION)
        }
    }
    
    // MARK: - Detection Methods
    
    // 1. Проверка отладки
    private fun isBeingDebugged(): Boolean {
        // Проверка через Debug.isDebuggerConnected()
        if (Debug.isDebuggerConnected()) {
            return true
        }
        
        // Проверка через TracerPid
        try {
            val reader = java.io.BufferedReader(
                java.io.FileReader("/proc/${Process.myPid()}/status")
            )
            var line: String?
            while (reader.readLine().also { line = it } != null) {
                if (line?.startsWith("TracerPid:") == true) {
                    val tracerPid = line?.substringAfter("TracerPid:").trim()?.toIntOrNull() ?: 0
                    if (tracerPid != 0) {
                        return true
                    }
                }
            }
            reader.close()
        } catch (e: IOException) {
            Log.e("RASP", "Error checking debugger status", e)
        }
        
        return false
    }
    
    // 2. Проверка модификации кода
    private fun isCodeModified(): Boolean {
        // Проверка подписи приложения
        try {
            val packageInfo = context.packageManager.getPackageInfo(
                context.packageName,
                android.content.pm.PackageManager.GET_SIGNATURES
            )
            
            val signatures = packageInfo.signatures
            if (signatures.isNotEmpty()) {
                val signature = signatures[0]
                val md = java.security.MessageDigest.getInstance("SHA")
                md.update(signature.toByteArray())
                val hash = md.digest()
                
                // В реальном приложении здесь была бы проверка с ожидаемым хешем
                return false
            }
        } catch (e: Exception) {
            Log.e("RASP", "Error checking code signature", e)
            return true
        }
        
        return false
    }
    
    // 3. Проверка инъекций
    private fun isCodeInjected(): Boolean {
        // Проверка на подозрительные библиотеки
        val suspiciousLibraries = listOf(
            "libsubstrate.so",
            "libhooker.so",
            "libsubstitute.so",
            "libcynject.so"
        )
        
        for (library in suspiciousLibraries) {
            if (isLibraryLoaded(library)) {
                return true
            }
        }
        
        return false
    }
    
    // 4. Проверка эмуляции
    private fun isRunningOnEmulator(): Boolean {
        // Проверка свойств эмулятора
        val emulatorProperties = mapOf(
            "ro.kernel.qemu" to "1",
            "ro.hardware" to "goldfish",
            "ro.product.model" to "sdk_gphone",
            "ro.build.product" to "sdk_gphone"
        )
        
        for ((property, value) in emulatorProperties) {
            if (getSystemProperty(property) == value) {
                return true
            }
        }
        
        return false
    }
    
    // 5. Проверка hooking
    private fun isHooked(): Boolean {
        // Проверка на подозрительные методы
        val suspiciousMethods = listOf(
            "hookMethod",
            "hookNativeMethod",
            "replaceMethod"
        )
        
        for (methodName in suspiciousMethods) {
            if (isMethodHooked(methodName)) {
                return true
            }
        }
        
        return false
    }
    
    // 6. Проверка памяти
    private fun isMemoryTampered(): Boolean {
        // Проверка на подозрительные паттерны в памяти
        val suspiciousPatterns = listOf(
            "HOOK",
            "INJECT",
            "PATCH"
        )
        
        for (pattern in suspiciousPatterns) {
            if (isPatternInMemory(pattern)) {
                return true
            }
        }
        
        return false
    }
    
    // 7. Проверка целостности
    private fun isIntegrityCompromised(): Boolean {
        // Проверка на изменения в критических классах
        val criticalClasses = listOf(
            "com.aladdin.mobile.security.CertificatePinningManager",
            "com.aladdin.mobile.security.RootDetector",
            "com.aladdin.mobile.network.ALADDINNetworkManager"
        )
        
        for (className in criticalClasses) {
            if (isClassModified(className)) {
                return true
            }
        }
        
        return false
    }
    
    // MARK: - Threat Handling
    
    private fun handleSecurityThreat(threat: SecurityThreat) {
        threatCount++
        lastThreatTime = System.currentTimeMillis()
        
        Log.w("RASP", "Threat Detected: ${threat.name}")
        
        // Выполняем защитные действия
        executeDefensiveActions(threat)
    }
    
    private fun executeDefensiveActions(threat: SecurityThreat) {
        when (threat) {
            SecurityThreat.DEBUGGING -> {
                // Завершаем приложение при отладке
                Process.killProcess(Process.myPid())
            }
            SecurityThreat.CODE_MODIFICATION -> {
                // Блокируем подозрительные функции
                disableSensitiveFeatures()
            }
            SecurityThreat.CODE_INJECTION -> {
                // Очищаем память
                clearSensitiveData()
            }
            SecurityThreat.EMULATION -> {
                // Ограничиваем функциональность
                limitFunctionality()
            }
            SecurityThreat.HOOKING -> {
                // Перезапускаем критические компоненты
                restartCriticalComponents()
            }
            SecurityThreat.MEMORY_TAMPERING -> {
                // Очищаем и перезагружаем данные
                reloadSecurityData()
            }
            SecurityThreat.INTEGRITY_VIOLATION -> {
                // Активируем максимальную защиту
                activateMaximumSecurity()
            }
        }
    }
    
    // MARK: - Defensive Actions
    
    private fun disableSensitiveFeatures() {
        Log.i("RASP", "Disabling sensitive features")
        // В реальном приложении здесь была бы логика отключения функций
    }
    
    private fun clearSensitiveData() {
        Log.i("RASP", "Clearing sensitive data")
        // В реальном приложении здесь была бы очистка данных
    }
    
    private fun limitFunctionality() {
        Log.i("RASP", "Limiting functionality")
        // В реальном приложении здесь было бы ограничение функций
    }
    
    private fun restartCriticalComponents() {
        Log.i("RASP", "Restarting critical components")
        // В реальном приложении здесь был бы перезапуск компонентов
    }
    
    private fun reloadSecurityData() {
        Log.i("RASP", "Reloading security data")
        // В реальном приложении здесь была бы перезагрузка данных
    }
    
    private fun activateMaximumSecurity() {
        Log.i("RASP", "Activating maximum security")
        // В реальном приложении здесь была бы активация максимальной защиты
    }
    
    // MARK: - Helper Methods
    
    private fun isLibraryLoaded(libraryName: String): Boolean {
        return try {
            System.loadLibrary(libraryName)
            true
        } catch (e: UnsatisfiedLinkError) {
            false
        }
    }
    
    private fun getSystemProperty(property: String): String {
        return try {
            val process = Runtime.getRuntime().exec("getprop $property")
            val reader = process.inputStream.bufferedReader()
            val result = reader.readLine() ?: ""
            reader.close()
            result
        } catch (e: Exception) {
            ""
        }
    }
    
    private fun isMethodHooked(methodName: String): Boolean {
        // В реальном приложении здесь была бы проверка на hooking
        return false
    }
    
    private fun isPatternInMemory(pattern: String): Boolean {
        // В реальном приложении здесь была бы проверка памяти
        return false
    }
    
    private fun isClassModified(className: String): Boolean {
        // В реальном приложении здесь была бы проверка целостности класса
        return false
    }
    
    private fun determineSecurityLevel(): SecurityLevel {
        return when {
            threatCount == 0 -> SecurityLevel.HIGH
            threatCount < 3 -> SecurityLevel.MEDIUM
            else -> SecurityLevel.LOW
        }
    }
    
    /**
     * Получить детальный отчет
     */
    fun getDetailedReport(): String {
        val status = getMonitoringStatus()
        
        var report = "🛡️ RASP Security Report\n"
        report += "======================\n"
        report += "Monitoring: ${if (status.isMonitoring) "✅ ACTIVE" else "❌ INACTIVE"}\n"
        report += "Threats Detected: ${status.threatCount}\n"
        report += "Last Threat: ${if (status.lastThreatTime > 0) java.util.Date(status.lastThreatTime) else "None"}\n"
        report += "Security Level: ${status.securityLevel.name}\n"
        
        return report
    }
}

// MARK: - Data Models

data class RASPStatus(
    val isMonitoring: Boolean,
    val threatCount: Int,
    val lastThreatTime: Long,
    val securityLevel: SecurityLevel
)

enum class SecurityThreat {
    DEBUGGING,
    CODE_MODIFICATION,
    CODE_INJECTION,
    EMULATION,
    HOOKING,
    MEMORY_TAMPERING,
    INTEGRITY_VIOLATION
}

enum class SecurityLevel {
    LOW, MEDIUM, HIGH
}

